### ISSUE TITLE
Injecting the appDelegate into Singletons creates a second AppDelegate

### ISSUE THREAD

**Issue by dmueller39** [appsquickly/Typhoon#449]
I assumed that I could inject the appDelegate into another object as its delegate:

```swift
dynamic func appDelegate() -> AnyObject {
        return TyphoonDefinition.withClass(AppDelegate.self) {
            (definition) in
            definition.injectProperty("assembly")
        }
}


dynamic func authenticationManager() -> AnyObject {
    return TyphoonDefinition.withClass(AuthenticationManager.self) {
            (definition) in
            definition.injectProperty("delegate", with: self.appDelegate())
            definition.scope = TyphoonScope.LazySingleton
    }
}
```

Doing this results in a second app delegate being created prior to `injectInitialFactoryIntoDelegate` getting called in `swizzleSetDelegateMethodOnApplicationClass` This second, unreferenced app delegate is then injected everywhere, and instantly dereferenced.

**Comment by igrekde** [comment]
Hi, @dmueller39. Looks like a bug, we'll investigate it ASAP. Thanks for reporting!

**Comment by igrekde** [comment]
@dmueller39, everything works fine when you assign the `AppDelegate` singletone scope:

```objc
dynamic func appDelegate() -> AnyObject {
        return TyphoonDefinition.withClass(AppDelegate.self) {
            (definition) in
            definition.injectProperty("assembly")
            definition.scope = TyphoonScope.Singletone
        }
}
```

Current behavior is fine, because the default scope is `ObjectGraph`. This scope is essential (and unique to Typhoon) for mobile and desktop applications. When a component is resolved, any dependencies with the object-graph will be treated as shared instances during resolution. Once resolution is complete they are not retained by the `TyphoonComponentFactory`. This allows instantiating an entire object graph for a use-case (say for a ViewController), and then discarding it when that use-case has completed, therefore making efficient use of memory.

In this situation you have two use-cases - one for instantiating the initial AppDelegate, and another - for injecting your `authenticationManager` in a view controller.

Does it help you?

**Comment by dmueller39** [comment]
Apply the attached patch (449.txt)  to https://github.com/appsquickly/Typhoon-example to see what I am talking about.

[449.txt](https://github.com/appsquickly/Typhoon/files/40522/449.txt)

**Comment by igrekde** [comment]
@dmueller39 Sorry, I was wrong. The setup from your first snippet seems to work fine in the demo app case:
```
- (PFAppDelegate *)appDelegate
{
    return [TyphoonDefinition withClass:[PFAppDelegate class] configuration:^(TyphoonDefinition *definition)
    {
        [definition injectProperty:@selector(window) with:[self mainWindow]];
        [definition injectProperty:@selector(cityDao) with:[_coreComponents cityDao]];
        [definition injectProperty:@selector(rootViewController) with:[self rootViewController]];
    }];
}

- (PFRootViewController *)rootViewController
{
    return [TyphoonDefinition withClass:[PFRootViewController class] configuration:^(TyphoonDefinition *definition)
    {
        [definition useInitializer:@selector(initWithMainContentViewController:assembly:) parameters:^(TyphoonMethod *initializer)
        {
            [initializer injectParameterWith:[self weatherReportController]];
            [initializer injectParameterWith:self];
        }];
        [definition injectProperty:@selector(delegate) with:[self appDelegate]];
        definition.scope = TyphoonScopeLazySingleton;
    }];
}
```

**Comment by dmueller39** [comment]
@igrekde I think I misunderstood the problem at first. It specifically occurs when definitions have the scope TyphoonScopeSingleton AND are injected with the appDelegate. I can see the additional AppDelegate being created in the call stack under `- (void)instantiateEagerSingletons`

![first_app_delegate](https://cloud.githubusercontent.com/assets/9422359/11337945/f5036a9c-91be-11e5-96bd-0ad6bbeaf3ba.png)
![second_app_delegate](https://cloud.githubusercontent.com/assets/9422359/11337948/f6f213d0-91be-11e5-9291-cfb9885ccee3.png)

I used this setup to trigger this.

```
- (PFAppDelegate *)appDelegate
{
    return [TyphoonDefinition withClass:[PFAppDelegate class] configuration:^(TyphoonDefinition *definition)
    {
        [definition injectProperty:@selector(window) with:[self mainWindow]];
        [definition injectProperty:@selector(cityDao) with:[_coreComponents cityDao]];
        [definition injectProperty:@selector(rootViewController) with:[self rootViewController]];
    }];
}

- (PFRootViewController *)rootViewController
{
    return [TyphoonDefinition withClass:[PFRootViewController class] configuration:^(TyphoonDefinition *definition)
    {
        [definition useInitializer:@selector(initWithMainContentViewController:assembly:) parameters:^(TyphoonMethod *initializer)
        {
            [initializer injectParameterWith:[self weatherReportController]];
            [initializer injectParameterWith:self];
        }];
        [definition injectProperty:@selector(delegate) with:[self appDelegate]];
        definition.scope = TyphoonScopeSingleton; //this is a singleton, and is instantiated as an eager singleton
    }];
}
```

I'm not saying this is a bug, I'm fine with using a lazy singleton, but the appDelegate seems to be a special case when it comes to building the object graph. This is unexpected based on what the documentation tells me. Either there should be a special note about AppDelegates in the documentation, or the framework should throw an exception somehow.

Here's a suggestion:
An explicit way to handle this would be to define a scope `TyphoonScopeExternal` or something else that's appropriate that throws an exception when the framework is attempting to instantiate a new one internally (which it never should directly, in theory).

**Comment by lemonkey** [comment]
While probably not related, one thing I've run into with an older iOS project that was created before Xcode 7 is that after upgrading to Xcode 7/iOS 9, the unit tests are exhibiting really strange behavior. 

It seems like the test target is being injected into the test host later than usual, such that the usual tricks people do to have one app delegate be used for normal execution while having a separate slimmed down version just for unit testing won't work anymore (by checking for the existence of an object of a specific class in main.m, etc.), among other issues related to swizzled methods that used to work during unit tests no longer do. Mind you this project is still using Typhoon 3.2.8, but is being updated to use the latest version soon.

I'm not sure if it's more of an issue of the project having been created before Xcode 7 or issues with Xcode 7, unit tests and Typhoon in particular.

**Comment by igrekde** [comment]
@dmueller39 Thanks for the detailed explanation!

I don't see the need of creating an exception in such a case - the `TyphoonLazySingleton` scope looks like the perfect solution in this case. The situation itself is not trivial, so I'm not sure that this has to be added in documentation - it's just the consequence of how `TyphoonScopeSingleton` objects are handled by `Typhoon`.

**Comment by igrekde** [comment]
@lemonkey Looks like a bug - it would be great if you were able to check this behavior with the latest version of Typhoon and post a separate issue. Thanks for reporting :)

**Comment by lemonkey** [comment]
@igrekde yeah going to look into it as part of my overall migration to dynamic frameworks for our shared library that uses Typhoon as a dependency. I'll let you know... :+1:

**Comment by lemonkey** [comment]
@igrekde I can confirm that this bug no longer exists when using the latest version of Typhoon and after converting my shared static library to a dynamic framework.

**Comment by igrekde** [comment]
@lemonkey Thanks for reporting! Feel free to post an issue if you encounter it once more.
